EPs/sep_data_spectrum.md

```diff
@@ -0,0 +1,108 @@
+# Add functionality for spectrum data
+
+
+author: Kyle Boone <!-- your name -->
+
+date-created: 2020 May 18 <!-- replace with the date you submit the APE -->
+
+date-last-revised: 2020 May 18 <!-- keep this up to date anytime something changes -->
+
+date-accepted: <!-- replace with accepted date -->
+
+status: Discussion  <!-- one of: Discussion, Accepted, Rejected -->
+
+
+## Abstract
+
+The goal of this proposal is to add functionality to sncosmo to handle spectrum data. We
+propose to add a SpectrumData class that represents a spectra taken of a supernova, and
+implement functionality to work with this data.
+
+
+## Detailed description
+
+Sncosmo is currently limited to working with photometric data. However, spectra are
+often obtained of supernovae, and it would be very useful if sncosmo could use this
+additional data. We propose to add functionality to sncosmo to work with spectra.
+
+There are several ways in which spectra can be used for supernova cosmology. First,
+spectra can provide additional constraints when fitting models to observations of a
+supernova. Models such as the SNEMO model that was recently added to sncosmo are
+designed to be able to fit individual spectra of each supernova rather than being
+limited to photometry. Surveys such as the SNfactory are obtaining spectrophotometric
+timeseries of supernovae that can then be fit directly with these models, potentially
+providing additional information beyond what is captured by photometry alone. 
+
+Given a spectrophotometric timeseries, it is possible to generate synthetic photometry
+in arbitrary bands. This photometry has many uses, such as comparing between experiments
+or using light curve fitters that take photometry as input. Sncosmo already provides
+functionality to generate photometry from spectral time series models, and this same
+functionality could be applied to spectrophotometric spectra.
+
+Less-well-calibrated spectra can also be useful for typing supernovae or constraining
+models. However, they often require some warping of the spectrum or removal of host
+galaxy contamination. This functionality is beyond the scope of this EP, but we would
+like to set up a framework to be able to include it in the future.
+
+
+## Branches and pull requests
+
+An initial discussion of this work can be found in sncosmo/sncosmo#262.
+
+sncosmo/sncosmo#265 has an implementation of fitting models to spectral time series.
+
+sncosmo/sncosmo#259 implements basic synthesized photometry.
+
+
+## Implementation
+
+The building block for all of this work is adding a SpectrumData class to sncosmo to
```

Comment:
I like the idea of renaming `Spectrum` to `ModelSpectrum` and calling this new class `Spectrum`. I wasn't sure about how to handle deprecation... the `Spectrum` class is internal and has never been exposed. The new `Spectrum` class will expose an almost identical API, so maybe that's OK even if anybody was actually using it?

Maybe we should try to check if there was anybody using the `Spectrum` class? I can ask around the SNfactory/WFIRST groups since they are probably the most likely ones.

Reply:
I think to do the deprecation we could:
- Copy `Spectrum` -> `ModelSpectrum`
- Change all internal references to use `ModelSpectrum`
- Put a deprecation message in `Spectrum.__init__` saying that the API of Spectrum will change in the next release and to use `ModelSpectrum` if you really need it (though it isn't recommended since it's not public).

Easy enough, but  it will delay implementation of this EP for a release. I'm not sure if that is worth the hassle or not. Were you able to determine if anyone is using it?

Reply:
It sounds like `Spectrum` is being used by several people right now for synthetic photometry on real spectra.

One option would be to do the renaming but still keep the same API. All that Spectrum does right now is compute bandfluxes, so we just have to keep compatible signatures for `__init__` and `bandflux` (which is currently the case). There would be minor differences in the computed bandfluxes for low-resolution spectra from interpolating vs explicitly handling the bin edges, but I would argue that the new approach is doing a better job for real data so it would be an improvement.

If we do that, could get away without deprecating anything?

Reply:
That sounds like a great approach, if it's possible! I think we wouldn't have to deprecate anything. Minor differences in the results of bandfluxes are fine, especially considering that this has been, up until now, an undocumented part of the API.